File: frontend/src/services/listService.js
// src/services/listService.js

import axios from 'axios';
import {getUserFavoritesIdFromStorage} from "./userService.js";

const API_URL = 'api/users';

export const addToList = async (userId, listId, itemId) => {
    const response = await axios.post(`${API_URL}/diner_user/${userId}/lists/${listId}/items/${itemId}`);
    return response.data;
}
export const addToFavorites = async (userId, itemId) => {
    const response = await axios.post(`${API_URL}/diner_user/${userId}/lists/${getUserFavoritesIdFromStorage()}/items/${itemId}`);
    return response.data;
}
export const removeFromList = async (userId, listId,itemId) => {
    await axios.delete(`${API_URL}/diner_user/${userId}/lists/${listId}/items/${itemId}`);
}
export const createList = async (userId, name, isPublic) => {

    const response = await axios.post(
        `${API_URL}/diner_user/${userId}/lists`, {
            name: name,
            "is_public": isPublic
        });
    return response.data;

}
export const removeList = async (userId, listId) => {

    await axios.delete(`${API_URL}/diner_user/${userId}/lists/${listId}`);
}

export const updateList = async (userId, name, isPublic, listId) => {
    const response = await axios.patch(
        `${API_URL}/diner_user/${userId}/lists/${listId}`,
        {
            name: name,
            "is_public": isPublic,
        });
    return response.data;
}



/** Kullanıcının kendi listelerini döner */
export const getUserLists = async (id) => {

    const listResponse = await axios.get(`${API_URL}/diner_user/${id}/lists`);
    const lists = listResponse.data;

    return lists;

}
export const getUserListItems = async(id,listId) => {
    const listItemResponse = await axios.get(`${API_URL}/diner_user/${id}/lists/${listId}/items`);

    return listItemResponse.data;
}

/** Halka açık listeleri döner */
export const getPublicLists = async () => {
    const listResponse = await axios.get(`${API_URL}/diner_user/public/lists`);
    return listResponse.data;

}

export const toggleFavorite = async (userId, itemId) => {
    const favorites = await getUserListItems(userId, getUserFavoritesIdFromStorage());
    const isFavorited = favorites.some(fav => fav.id === itemId);

    if (isFavorited) {
        await removeFromList(userId, getUserFavoritesIdFromStorage(), itemId);
    } else {
        await addToFavorites(userId, itemId);
    }

    return await getUserListItems(userId, getUserFavoritesIdFromStorage());
}

File: IOS/Services/ListService.swift
import Foundation

/// Service handling list operations through the API client.
final class ListService: @unchecked Sendable {
    private let api: APIClientProtocol
    private let userService: UserService

    init(api: APIClientProtocol = APIClient.shared, userService: UserService = UserService()) {
        self.api = api
        self.userService = userService
    }

    private let base = "api/users/diner_user"

    func addToList(userId: String, listId: String, itemId: String) async throws -> EmptyResponse {
        let path = "\(base)/\(userId)/lists/\(listId)/items/\(itemId)"
        return try await api.request(path, method: "POST", body: nil, useCache: false, cacheTTL: nil)
    }

    func addToFavorites(userId: String, itemId: String) async throws -> EmptyResponse {
        guard let favoritesId = userService.getUserFavoritesIdFromStorage() else { return EmptyResponse() }
        return try await addToList(userId: userId, listId: favoritesId, itemId: itemId)
    }

    func removeFromList(userId: String, listId: String, itemId: String) async throws -> EmptyResponse {
        let path = "\(base)/\(userId)/lists/\(listId)/items/\(itemId)"
        return try await api.request(path, method: "DELETE", body: nil, useCache: false, cacheTTL: nil)
    }

    func createList(userId: String, name: String, isPublic: Bool) async throws -> UserList {
        let body = try JSONEncoder().encode(ListRequest(name: name, isPublic: isPublic))
        let path = "\(base)/\(userId)/lists"
        return try await api.request(path, method: "POST", body: body, useCache: false, cacheTTL: nil)
    }

    func removeList(userId: String, listId: String) async throws -> EmptyResponse {
        let path = "\(base)/\(userId)/lists/\(listId)"
        return try await api.request(path, method: "DELETE", body: nil, useCache: false, cacheTTL: nil)
    }

    func updateList(userId: String, listId: String, name: String, isPublic: Bool) async throws -> UserList {
        let body = try JSONEncoder().encode(ListRequest(name: name, isPublic: isPublic))
        let path = "\(base)/\(userId)/lists/\(listId)"
        return try await api.request(path, method: "PATCH", body: body, useCache: false, cacheTTL: nil)
    }

    func getUserLists(userId: String) async throws -> [UserList] {
        let path = "\(base)/\(userId)/lists"
        // Cache user lists for 5 minutes
        return try await api.request(path, method: "GET", body: nil, useCache: true, cacheTTL: 300)
    }

    func getUserListItems(userId: String, listId: String) async throws -> [Restaurant] {
        let path = "\(base)/\(userId)/lists/\(listId)/items"
        // Cache list items for 3 minutes - favorites might change
        let dtos: [BusinessDTO] = try await api.request(path, method: "GET", body: nil, useCache: true, cacheTTL: 180)
        return dtos.map(BusinessMapper.map)
    }

    func getPublicLists() async throws -> [UserList] {
        let path = "\(base)/public/lists"
        // Cache public lists for 10 minutes
        return try await api.request(path, method: "GET", body: nil, useCache: true, cacheTTL: 600)
    }

    func toggleFavorite(userId: String, itemId: String) async throws -> [Restaurant] {
        guard let favoritesId = userService.getUserFavoritesIdFromStorage() else { return [] }
        let favorites = try await getUserListItems(userId: userId, listId: favoritesId)
        let isFavorited = favorites.contains { $0.id == itemId }

        if isFavorited {
            _ = try await removeFromList(userId: userId, listId: favoritesId, itemId: itemId)
        } else {
            _ = try await addToFavorites(userId: userId, itemId: itemId)
        }

        return try await getUserListItems(userId: userId, listId: favoritesId)
    }
}


private struct ListRequest: Encodable {
    let name: String
    let isPublic: Bool

    enum CodingKeys: String, CodingKey {
        case name
        case isPublic = "is_public"
    }
}
